### ISSUE TITLE
[ROLE PARITY] Determine which of these elements with existing ARIA role mappings require a new role

### ISSUE THREAD

**Issue by joanmarie** [w3c/aria#696]
_N.B. All of the "Role Parity" issues are intended to gather feedback from stakeholders who need role parity from us including, but not limited to, groups working on Web Components and AOM. Comments on matters other than determining "which of these elements with existing ARIA role mappings require a new role" should be raised in another GitHub issue. Thank you for your understanding!_

The following elements are already mapped in the HTML AAM to an ARIA role. In addition, the HTML AAM platform mapping for all platforms is "Use WAI-ARIA mapping." HOWEVER, unlike the elements listed in https://github.com/w3c/aria/issues/694, there is not a unique correspondence between the element and the ARIA role. For instance, both `dfn` and `dt` map to the ARIA `term` role.

Assumption: Some of these elements may need a dedicated role; others may not. For instance, what makes it clear that an `ol` is a `list` which is ordered is the presence of the indexed children. Thus those elements might already have sufficient "parity."

We need to determine which of the items below belong in the group of elements in https://github.com/w3c/aria/issues/694, which do not need a new role but _might_ benefit from a new attribute (e.g. the type of button to distinguish `submit` from `reset`), and which _definitely_ need a new role.

* [`a` (represents a hyperlink)](https://w3c.github.io/html-aam/#el-a): [`link`](https://rawgit.com/w3c/core-aam/master/#role-map-link) - OK
* [`a` (represents a hyperlink and parent is a menu)](https://w3c.github.io/html-aam/#el-a-parentmenu): [`menuitem`](https://rawgit.com/w3c/core-aam/master/#role-map-menuitem) - Patches welcome
* [`area` (represents a hyperlink)](https://w3c.github.io/html-aam/#el-area): [`link`](https://rawgit.com/w3c/core-aam/master/#role-map-link) - OK
* [`body`](https://w3c.github.io/html-aam/#el-body): [document](https://rawgit.com/w3c/core-aam/master/#role-map-document) (has [issue](https://github.com/w3c/html-aam/issues/117)) - no parity will be attempted
* [`datalist` (represents pre-defined options for `input` element)](https://w3c.github.io/html-aam/#el-datalist): [`listbox`](https://rawgit.com/w3c/core-aam/master/#role-map-listbox) - OK
* [`dd`](https://w3c.github.io/html-aam/#el-dd): [`definition`](https://rawgit.com/w3c/core-aam/master/#role-map-definition) - OK
* [`dfn`](https://w3c.github.io/html-aam/#el-dfn): [`term`](https://rawgit.com/w3c/core-aam/master/#role-map-term) - generic `inline` with TBD attribute
* [`dt`](https://w3c.github.io/html-aam/#el-dt): [`term`](https://rawgit.com/w3c/core-aam/master/#role-map-term) - no new role, but if in definition list role will support aria-posinset and aria-setsize
* [`img` (empty alt value)](https://w3c.github.io/html-aam/#el-img-emptyalt): [`none`](https://rawgit.com/w3c/core-aam/master/#role-map-none) (A.K.A. [`presentation`](https://rawgit.com/w3c/core-aam/master/#role-map-presentation)) - OK
* [`input` (`type=button`, parent is menu)](https://w3c.github.io/html-aam/#el-input-button-parentmenu): [`menuitem`](https://rawgit.com/w3c/core-aam/master/#role-map-menuitem) Patches welcome
* [`input` (`type=checkboxstate`, parent is menu)](https://w3c.github.io/html-aam/#el-input-checkbox-parentmenu): [`menuitemcheckbox`](https://rawgit.com/w3c/core-aam/master/#role-map-menuitemcheckbox) - Patches welcome
* [`input` (`type=email`, no suggestions source element)](https://w3c.github.io/html-aam/#el-input-email): [`textbox`](https://rawgit.com/w3c/core-aam/master/#role-map-textbox) - OK, but TBD attribute
* [`input` (`type=email,search,telephone,text,url`, suggestion source element)](https://w3c.github.io/html-aam/#el-input-textetc-autocomplete): [`combobox`](https://rawgit.com/w3c/core-aam/master/#role-map-combobox) - OK, but TBD attribute
* [`input` (`type=image`)](https://w3c.github.io/html-aam/#el-input-image): [`button`](https://rawgit.com/w3c/core-aam/master/#role-map-button) - OK
* [`input` (`type=image`, parent is menu)](https://w3c.github.io/html-aam/#el-input-image-parentmenu): [`menuitem`](https://rawgit.com/w3c/core-aam/master/#role-map-menuitem) Patches welcome
* [`input` (`type=radio`, parent is menu)](https://w3c.github.io/html-aam/#el-input-radio-menuparent) : [`menuitemradio`](https://rawgit.com/w3c/core-aam/master/#role-map-menuitemradio) - Patches welcome
* [`input` (`type=reset`)](https://w3c.github.io/html-aam/#el-input-reset): [`button`](https://rawgit.com/w3c/core-aam/master/#role-map-button) - OK
* [`input` (`type=submit`)](https://w3c.github.io/html-aam/#el-input_submit): [`button`](https://rawgit.com/w3c/core-aam/master/#role-map-button) - OK
* [`input` (`type=telephone`, no suggestions source element)](https://w3c.github.io/html-aam/#el-input-tel): [`textbox`](https://rawgit.com/w3c/core-aam/master/#role-map-textbox) - OK, but TBD attribute
* [`input` (`type=text`, no suggestions source element)](https://w3c.github.io/html-aam/#el-input-text): [`textbox`](https://rawgit.com/w3c/core-aam/master/#role-map-textbox) - OK, but TBD attribute
* [`input` (`type=url`, no suggestions source element)](https://w3c.github.io/html-aam/#el-input-url): [`textbox`](https://rawgit.com/w3c/core-aam/master/#role-map-textbox) - OK, but TBD attribute
* [`keygen`](https://w3c.github.io/html-aam/#el-keygen): [`listbox`](https://rawgit.com/w3c/core-aam/master/#role-map-listbox) - OK (`keygen` is deprecated)
* [`ol`](https://w3c.github.io/html-aam/#el-ol): [`list`](https://rawgit.com/w3c/core-aam/master/#role-map-list) - single role, but TBD attribute
* [`optgroup`](https://w3c.github.io/html-aam/#el-optgroup): [`group`](https://rawgit.com/w3c/core-aam/master/#role-map-group) - OK
* [`section` (with accessible name)](https://w3c.github.io/html-aam/#el-section): [`region`](https://rawgit.com/w3c/core-aam/master/#role-map-region) - OK
* [`select` (with `multiple` attribute or size > 1)](https://w3c.github.io/html-aam/#el-select-listbox): [`listbox`](https://rawgit.com/w3c/core-aam/master/#role-map-listbox)
* [`select` (no `multiple` attribute, size <= 1)](https://w3c.github.io/html-aam/#el-select-combobox): [`combobox`](https://rawgit.com/w3c/core-aam/master/#role-map-combobox) - change HTML-AAM mapping to `listbox` with `aria-expanded` set to `false` (needs to be added to `listbox`) and add `aria-valuetext` as a supported property of `listbox`
* [`tbody`](https://w3c.github.io/html-aam/#el-tbody): [`rowgroup`](https://rawgit.com/w3c/core-aam/master/#role-map-rowgroup)
* [`textarea`](https://w3c.github.io/html-aam/#el-textarea): [`textbox`, with mapping for aria-multiline=`true`](https://rawgit.com/w3c/core-aam/master/#role-map-textbox-multiline)
* [`thead`](https://w3c.github.io/html-aam/#el-thead): [`rowgroup`](https://rawgit.com/w3c/core-aam/master/#role-map-rowgroup) - OK, but possibly attribute
* [`tfoot`](https://w3c.github.io/html-aam/#el-tfoot): [`rowgroup`](https://rawgit.com/w3c/core-aam/master/#role-map-rowgroup) - OK, but possibly attribute
* [`ul`](https://w3c.github.io/html-aam/#el-ul): [`list`](https://rawgit.com/w3c/core-aam/master/#role-map-list) - single role, but TBD attribute - OK, but possibly attribute

**Comment by cookiecrook** [comment]
FWIW, it's good to reference the `presentation` role as the newer synonym `none` wherever possible. I made one minor edit to you issue summary above.

**Comment by carmacleod** [comment]
Question about `section (with accessible name): region`:

Does `accessible name` include (or will it ever include) the first heading child (`h1 - h6 or even h`), even without an explicit `aria-labelledby` pointing to that child?

**Comment by mcking65** [comment]
carmacleod commented:

> Question about section (with accessible name): region:
>
> Does accessible name include (or will it ever include) the first heading child (h1 - h6 or even h), even without an explicit aria-labelledby pointing to that child?

While I understand the appeal, I would not support it.

1. Browsers should not guess at author intentions with respect to content. Semantics should be explicit.
2. I do not think it would be good to make such a feature an explicit part of the HTML spec; it would be unnecessarily limiting for authors.
3. In current content, It would create lots of regions where regions should not exist; not every section should be a region, even if that section contains a heading.

**Comment by carmacleod** [comment]
Agree completely, @mcking65.

I was asking because I hoped the answer was "no".   :)
After mulling over what you said in https://github.com/w3c/aria-practices/issues/562, I finally "get it":
- Outlines are for the information architecture of the page, aka "the content".
- Landmarks are a convenient way to skip around the "structure" of the page.
- Sections with headings go in the Outline.
- Sections with aria-label or aria-labelledby are Landmarks.
- Landmarks don't belong in the Outline unless they have a heading child with `aria-labelledby="id-of-heading-child"`.
- Landmark navigation and Outline navigation are completely separate concepts, and they always will be, even if "the broken outline algorithm" is fixed some day.

So I think we should explicitly state that the `accessible name` for a section element can _only_ be defined by `aria-label` or `aria-labelledby`. (i.e. Don't leave it open to interpretation).

Further, we should make it explicitly clear that a heading child that is not referred to by aria-labelledby does not, and will never, define an `accessible name` for a section element. This means that a heading child cannot by itself change the semantics of a section from "thing that belongs in an outline" to "landmark region". Only `aria-label` or `aria-labelledby` can do that.

I think that if we make this abundantly clear in all specs, then the world can move forward on the outline vs landmarks stalemate.

The following are sources of confusion that need clarification:
1) The [HTML spec](http://w3c.github.io/html/sections.html#the-section-element) says that region is the default role for section, however it does not say that:
2) Setting `role="region"` on an element does not make it a [landmark] region unless it has an aria-label or aria-labelledby.
3) The HTML spec for section has other statements that could also be misleading, such as:
> - "Each section should be identified, typically by including a heading (h1-h6 element) as a child of the section element." ("identifying" could be mistaken for "giving an accessible name")
> - "A general rule is that the section element is appropriate only if the element’s contents would be listed explicitly in the document’s outline." (... it's not "only" for outlines... it's also for defining a landmark if it has an accessible name)
> - "For example the role of the element, which in this case is "region", can be announced by screen reader software when a user navigates to an section element." (.. not if it doesn't have a name...)
> - "When an element is needed only for styling purposes or as a convenience for scripting, authors are encouraged to use the div element instead." (this note needs to have stronger words, like using section without either a heading or an accessible name is an author error)

Hopefully we can help tighten all of these statements up.

(Sorry to dump this all in a "Role Parity" issue. I can create a new issue if this is off topic. I just wanted to have the discussion here, so that I can create a pull request for the HTML spec clarifications.)

**Comment by joanmarie** [comment]
@carmacleod: regarding this:

> (Sorry to dump this all in a "Role Parity" issue. I can create a new issue if this is off topic. I just wanted to have the discussion here, so that I can create a pull request for the HTML spec clarifications.)

Please do. All of the "Role Parity" issues are intended to gather feedback from stakeholders who need role parity from us, such as the folks working on Web Components and AOM. These groups are the main "customers" of role parity, and we don't know for certain what it is they need from us. The questions you raise here are certainly important, but as best as I can tell are largely irrelevant to what we are soliciting. If I am wrong about that, my apologies.

**Comment by domenic** [comment]
For web components purposes, what we are interested in is being able to create custom elements which present themselves to AT in the same way as these native elements. So for our purposes, the question is, e.g.: do ATs expose the difference between `dfn` and `dt`?

If they do, then we also need a way to do so in web components (whether it be a new attribute or new role).

If they do not distinguish, then we do not need any new roles or attributes.

Similarly, e.g. do ATs announce "ordered list" vs. "unordered list" or similar? If so, we need to be able to induce that behavior for `<custom-ordered-list>` and `<custom-unordered-list>` web components. But if they just announce "list", then the existing `list` role seems fine.

I'm unclear from the OP's summary which is the case, and am hopeful that those with domain expertise can weigh in to clarify.

**Comment by cookiecrook** [comment]
> For web components purposes, what we are interested in is being able to create custom elements which present themselves to AT in the same way as these native elements. 

Have you considered how you'd you make a custom element be operable like the HTML <video> element? In other words, could a custom element opt into exposing APIs like play(), pause(), etc?

**Comment by domenic** [comment]
That seems unrelated to this thread, perhaps best filed on the web components repo? But yes, my understanding is that AOM was meant to help with that.

**Comment by joanmarie** [comment]
@domenic: A given AT can do whatever it sees fit, of course. Something I'll come back to momentarily. But to hopefully clarify things regarding questions like:

> So for our purposes, the question is, e.g.: do ATs expose the difference between dfn and dt?

And:

> Similarly, e.g. do ATs announce "ordered list" vs. "unordered list" or similar? 

TL;DR: Maybe.

Long version:

If you look at the mappings in the HTML AAM (for which links have just been added above), you'll notice that the mappings are almost entirely "Use WAI-ARIA mapping." That means see what the Core AAM says by following the link on the right. In the case of `dfn` and `dt`, each of which map to `term`, you'll find in the [Core AAM](https://rawgit.com/w3c/core-aam/master/#role-map-term):

<table>
  <tr>
    <th>MSAA + IAccessible2</th>
    <td>Role: <code>IA2_ROLE_TEXT_FRAME</code><br>Object Attribute: <code>xml-roles:term</code></td>
  </tr>
  <tr>
    <th>UIA</th>
    <td>Control Type: <code>Text</code><br>Localized Control Type: <code>term</code></td>
  </tr>
  <tr>
    <th>ATK/AT-SPI</th>
    <td>Role: <code>ROLE_DESCRIPTION_TERM</code></td>
  </tr>
  <tr>
    <th>AX API</th>
    <td>AXRole: <code>AXGroup</code><br>AXSubrole: <code>AXTerm</code><br>AXRoleDescription: <code>'term'</code></td>
  </tr>
</table>

What the Core AAM mappings are is what user agents expose to assistive technologies via platform accessibility APIs for ARIA roles. As you will see in the table above, there is nothing _in the stated mapping_ that let's an AT determine (and present to the AT user) whether the element is a `dfn` or `dt`. You'll find the same scenario for `ol` and `ul` with [`list`](https://rawgit.com/w3c/core-aam/master/#role-map-list).

That said, there might be ways for ATs to still make that distinction. For instance, Gecko exposes the tag via ATK and IAccessible2. So an AT on those platforms _could_ present a distinction should their developers be so motivated. My guess is that all platforms and user agents have little things like that.... And, yes, of course that leads to inconsistent user experiences.... That's another matter.

Getting back to your questions, I _suspect_ my answer is more information than you were really wanting... :woman_shrugging: BUT, your answers suggest to me what the next step is, namely for the ARIA Working Group to get input from the AT developers regarding where _they_ want/need the ability to distinguish elements. See for instance what I said [here](https://github.com/w3c/aria/issues/698#issuecomment-363537503) regarding `sub` and `sup`.

Thanks again very much for helping us understand what you need -- and for helping us see what else we need to do. :smile:

**Comment by joanmarie** [comment]
Please note that the three deleted comments were copied and pasted in their entirety in https://github.com/w3c/aria-practices/issues/575.

**Comment by annevk** [comment]
> And, yes, of course that leads to inconsistent user experiences.... That's another matter.

Flushing that out more would be good though. E.g., is `tag` (sic) something everyone should expose or is it something that perhaps should go away? If it's important, is it okay for custom elements to "spoof" the value or does it need to be the custom element name, even if that leads to a different experience than intended?

**Comment by joanmarie** [comment]
This is a meta issue. The 1.3 TODO's can be found on https://github.com/w3c/aria/wiki/Plans-regarding-role-parity
